docs/running_lbann.rst

```diff
@@ -1,96 +1,434 @@
 .. role:: bash(code)
           :language: bash
+.. role:: python(code)
+          :language: python
 
-====================
+============================================================
 Running LBANN
-====================
+============================================================
 
-The basic template for running LBANN is
+------------------------------------------------
+Anatomy of an LBANN experiment
+------------------------------------------------
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Parallelism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+LBANN is run under the `MPI
+<https://en.wikipedia.org/wiki/Message_Passing_Interface>`_ paradigm,
+i.e. with multiple processes that communicate with message
+passing. These processes are subdivided into "trainers." Conceptually,
+a trainer owns parallel objects, like models and data readers, and
+generally operates independently of other trainers.
+
+Comments:
+
++ LBANN targets HPC systems with homogeneous compute nodes and GPU
+  accelerators, which motivates some simplifying assumptions:
+
+  - All trainers have the same number of processes.
+
+  - Each MPI process corresponds to one GPU.
+
++ Processors are block assigned to trainers based on MPI rank.
+
+  - In order to minimize the cost of intra-trainer communication, make
+    sure to map processes to the hardware and network
+    topologies. Typically, this just means choosing a sensible number
+    of processes per trainer, e.g. a multiple of the number of GPUs
+    per compute node.
+
++ Generally, increasing the number of processes per trainer will
```

Comment:
You may wish to note that this is very dependent on mini-batch size (and, more generally, partitioning scheme).